File: DDA.cpp
// ddaline.cpp

#include <iostream>
#include <vector>
#include <sstream>

void drawDDALine(int x1, int y1, int x2, int y2, std::vector<std::pair<int, int>> &result) {
    int dx = x2 - x1;
    int dy = y2 - y1;

    int steps = abs(dx) > abs(dy) ? abs(dx) : abs(dy);

    float xIncrement = static_cast<float>(dx) / static_cast<float>(steps);
    float yIncrement = static_cast<float>(dy) / static_cast<float>(steps);

    float x = static_cast<float>(x1), y = static_cast<float>(y1);

    // Store the coordinates in the result vector
    for (int k = 0; k < steps; k++) {
        result.push_back({static_cast<int>(x), static_cast<int>(y)});
        x += xIncrement;
        y += yIncrement;
    }
}

int main(int argc, char *argv[]) {
    if (argc != 5) {
        std::cerr << "Usage: " << argv[0] << " x1 y1 x2 y2" << std::endl;
        return 1;
    }

    int x1 = std::stoi(argv[1]);
    int y1 = std::stoi(argv[2]);
    int x2 = std::stoi(argv[3]);
    int y2 = std::stoi(argv[4]);

    std::vector<std::pair<int, int>> result;
    drawDDALine(x1, y1, x2, y2, result);

    // Print the resulting coordinates
    for (const auto &coord : result) {
        std::cout << "(" << coord.first << ", " << coord.second << ") ";
    }

    return 0;
}

File: DDA.js
// ddaline.js

function drawDDALine(x1, y1, x2, y2, result) {
    let dx = x2 - x1;
    let dy = y2 - y1;

    let steps = Math.max(Math.abs(dx), Math.abs(dy));

    let xIncrement = dx / steps;
    let yIncrement = dy / steps;

    let x = x1,
        y = y1;

    // Store the coordinates in the result array
    for (let k = 0; k < steps; k++) {
        result.push([Math.round(x), Math.round(y)]);
        x += xIncrement;
        y += yIncrement;
    }
}

function main() {
    const args = process.argv.slice(2);

    if (args.length !== 4) {
        console.error("Usage: node DDA.js x1 y1 x2 y2");
        process.exit(1);
    }

    const x1 = parseInt(args[0]);
    const y1 = parseInt(args[1]);
    const x2 = parseInt(args[2]);
    const y2 = parseInt(args[3]);

    const result = [];
    drawDDALine(x1, y1, x2, y2, result);

    // Output the coords to an external file in a format readable by matplotlib
    const fs = require('fs');
    const outputFile = fs.createWriteStream('output/DDA_coordinates.txt');
    result.forEach(coord => {
        outputFile.write(`${coord[0]} ${coord[1]}\n`);
    });
    outputFile.close();

    process.exit(0);
}

main();